transport-native-epoll/src/main/c/netty_epoll_native.c

```diff
@@ -103,7 +104,49 @@ jfieldID packetPortFieldId = NULL;
 jfieldID packetMemoryAddressFieldId = NULL;
 jfieldID packetCountFieldId = NULL;
 
+clockid_t epollWaitClock = 0; // initialized in initializeEpollWaitClock
+
 // util methods
+static jboolean initializeEpollWaitClock() {
+  struct timespec ts;
+  // First try to get a monotonic clock, as we effectively measure execution time and don't want the underlying clock
+  // moving unexpectedly/abruptly.
+  #ifdef CLOCK_MONOTONIC_COARSE
+  epollWaitClock = CLOCK_MONOTONIC_COARSE;
+  if (clock_gettime(epollWaitClock, &ts) != EINVAL) {
+    return JNI_TRUE;
+  }
+  #endif
+  #ifdef CLOCK_MONOTONIC_RAW
+  epollWaitClock = CLOCK_MONOTONIC_RAW;
+  if (clock_gettime(epollWaitClock, &ts) != EINVAL) {
+    return JNI_TRUE;
+  }
+  #endif
+  #ifdef CLOCK_MONOTONIC
+  epollWaitClock = CLOCK_MONOTONIC;
+  if (clock_gettime(epollWaitClock, &ts) != EINVAL) {
+    return JNI_TRUE;
+  }
+  #endif
+
+  // Fallback to realtime ... in this case we are subject to clock movements on the system.
+  #ifdef CLOCK_REALTIME_COARSE
+  epollWaitClock = CLOCK_REALTIME_COARSE;
+  if (clock_gettime(epollWaitClock, &ts) != EINVAL) {
+    return JNI_TRUE;
+  }
+  #endif
+  #ifdef CLOCK_REALTIME
+  epollWaitClock = CLOCK_REALTIME;
+  if (clock_gettime(epollWaitClock, &ts) != EINVAL) {
+    return JNI_TRUE;
+  }
+  #endif
+  fprintf(stderr, "FATAL: could not find a clock for clock_gettime!\n");
```

Comment:
0 may not be a legitimate value. It turns out to be defined as `CLOCK_REALTIME` on my system (which should be implemented everywhere according to [1]) so I think if we get to this point we are in a pretty bad state and should fail loudly.

[1] http://linux.die.net/man/3/clock_gettime
> All implementations support the system-wide realtime clock, which is identified by CLOCK_REALTIME